### PR TITLE
Add option to hide OBS windows from display capture

### PR DESCRIPTION
**Pull request by notr1ch**
### Description
This PR adds a new setting to hide the OBS application window and sub-windows / dialog boxes from screen capture applications (including OBS itself). This unfortunately does not cover 100% of the different types of windows that could appear, for example, context menus and tooltips, as the underlying API does not work well on these types of windows, causing a black box or black lines to appear on the captured output.

Only Windows 10 2004 and later is supported as that's when the `WDA_EXCLUDEFROMCAPTURE` parameter was added to [SetDisplayAffinity](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowdisplayaffinity). Cross-platform support is simple to add if the underlying APIs exist to be able to hide specific windows from capture APIs.

Projectors are special-cased and this option is not applied to them, as they are often intended to be visible and capturable by other applications.

### Motivation and Context
Users with a single monitor often have trouble with video feedback when using display capture. This option allows single monitor users to be able to manipulate the OBS UI without having OBS itself appear in the preview or output.

Demo:

https://user-images.githubusercontent.com/876345/147421792-4811842e-c472-4de9-9bac-2a75a45de5ae.mp4

### How Has This Been Tested?
Lots of testing on Windows 10 21H1. Tested behavior of the main window, dialog boxes, Qt message boxes, various sub-window, docks, projectors and browser docks. Further testing would be appreciated, PR is marked as draft for this reason.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: UI/obs-app.cpp [hunk 1]
```diff
@@ -1569,6 +1569,43 @@ bool OBSApp::TranslateString(const char *lookupVal, const char **out) const
 	return text_lookup_getstr(App()->GetTextLookup(), lookupVal, out);
 }
 
+// Global handler to receive all QEvent::Show events so	we can apply
+// display affinity on any newly created windows and dialogs without
+// caring where they are coming from (eg plugins).
```
Comment (RytoEX) [inline UI/obs-app.cpp:1574]: "eg" should be "e.g.,".

File: UI/window-basic-settings.cpp [hunk 2]
```diff
@@ -383,6 +383,7 @@ OBSBasicSettings::OBSBasicSettings(QWidget *parent)
 	HookWidget(ui->theme, 		     COMBO_CHANGED,  GENERAL_CHANGED);
 	HookWidget(ui->enableAutoUpdates,    CHECK_CHANGED,  GENERAL_CHANGED);
 	HookWidget(ui->openStatsOnStartup,   CHECK_CHANGED,  GENERAL_CHANGED);
+	HookWidget(ui->hideOBSFromCapture   ,CHECK_CHANGED,  GENERAL_CHANGED);
```
Comment (RytoEX) [inline UI/window-basic-settings.cpp:386]: Fix comma placement.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Minor nitpicks.

File: UI/data/locale/en-US.ini [hunk 3]
```diff
@@ -779,6 +779,7 @@ Basic.Settings.General.Theme="Theme"
 Basic.Settings.General.Language="Language"
 Basic.Settings.General.EnableAutoUpdates="Automatically check for updates on startup"
 Basic.Settings.General.OpenStatsOnStartup="Open stats dialog on startup"
+Basic.Settings.General.HideOBSWindowsFromCapture="Hide OBS windows from display capture"
```
[the remark this answers is not in the document]
Comment (notr1ch) [reply UI/data/locale/en-US.ini:782]: This affects all capture programs, not just OBS, so I intended it to be more generic.

[the remark this answers is not in the document]
Comment (RytoEX) [reply UI/data/locale/en-US.ini:782]: Ah, I see.  Then this is fine as-is.

**Comment by flaeri** [comment]
I really like it, and it seems to work really well for me, regardless of monitors and capture method.

Like you've noted, certain context menu / right click will show in preview and output, but no big deal imo. Big QOL add, and I would personally prefer it to be enabled by default.

I poked at it a fair amount, but couldn't really get it to misbehave in any way. Works well on my machine.
i7-8700, GTX 1070, Win 10 20H2

**Comment by WizardCM** [comment]
Verified this works for most of the main windows. Also verified it doesn't work for context menus & dropdown menus, which is expected.

Fullscreen projectors work fine, as do VST windows.

However, it doesn't work for windowed projectors.

**Comment by notr1ch** [comment]
> However, it doesn't work for windowed projectors. They still seem to be visible.

Projectors are intentionally excluded as there are cases of users capturing projectors with other apps, e.g. to show a PowerPoint presentation.

**Comment by Fenrirthviti** [comment]
Just wanted to throw in feedback if we can confirm this works reliable and has no detrimental effects, I would prefer it was enabled by default as well.

**Comment by notr1ch** [comment]
Pushed a new version with the changes requested above as well as new projector logic. Projectors now set a Qt property on the window handle so that they can easily be detected when the event handler finds a new window. This fixes the issue where fullscreen projectors also had their window affinity set to not capturable, since `showFullScreen()` was called before they were added to `allProjectors`, and it's nicer code overall.

**Pull request merged**